File: src/utils/gameLogic/timeline.ts
import { GAME_CONFIG } from '@/constants/gameConfig';
import type { GameState, Square } from '@/types/game';

import { isPlayingState } from './helpers';
import {
  detectPatterns,
  getPatternsByLeftColumn,
  markColumnCells,
} from './patterns';
import { clearMarkedCellsAndApplyGravity } from './physics';

/**
 * Update timeline sweep progression with column-based clearing logic
 */
export function updateTimeline(state: GameState): GameState {
  if (!isPlayingState(state)) {
    return state;
  }

  const newTimer = state.timeline.timer + 1;

  // Check if it's time to move timeline one column
  if (newTimer >= state.timeline.sweepInterval) {
    return advanceTimelineToNextColumn(state);
  }

  // Just increment timer
  return {
    ...state,
    timeline: {
      ...state.timeline,
      timer: newTimer,
    },
  };
}

/**
 * Advance timeline to next column and process the current column
 */
export function advanceTimelineToNextColumn(state: GameState): GameState {
  const currentColumn = state.timeline.x;
  const nextColumn = (currentColumn + 1) % GAME_CONFIG.board.width;

  // Process the current column before moving
  const processedState = processTimelineColumn(state, nextColumn);

  return {
    ...processedState,
    timeline: {
      ...processedState.timeline,
      x: nextColumn,
      timer: 0, // Reset timer for next column
    },
  };
}

/**
 * Process a single column when timeline passes through it
 */
export function processTimelineColumn(
  state: GameState,
  column: number
): GameState {
  const patternsInColumn = getPatternsByLeftColumn(
    state.detectedPatterns,
    column
  );
  const patternsInPreviousColumn = getPatternsByLeftColumn(
    state.detectedPatterns,
    column - 1
  );

  const hasPatternsInCurrentColumn = patternsInColumn.length > 0;
  const hasPatternsInPreviousColumn = patternsInPreviousColumn.length > 0;

  // Case 1: Mark cells for clearing if there are patterns
  if (hasPatternsInCurrentColumn || hasPatternsInPreviousColumn) {
    return markCellsForClearing(state, column, patternsInColumn);
  }

  // Case 2: Clear marked cells if no patterns and we have holding score
  const shouldClear =
    !hasPatternsInCurrentColumn &&
    !hasPatternsInPreviousColumn &&
    state.timeline.holdingScore > 0 &&
    state.markedCells.length > 0;

  if (shouldClear) {
    return clearMarkedCellsAndScore(state);
  }

  return state;
}

/**
 * Mark cells for clearing and update holding score
 */
export function markCellsForClearing(
  state: GameState,
  column: number,
  patternsInColumn: Square[]
): GameState {
  // Calculate holding points for patterns in current column
  const holdingPoints = patternsInColumn.length;

  // Mark cells in this column for clearing
  const newMarkedCells = markColumnCells(column, state.detectedPatterns);

  // Single state update with all changes
  return {
    ...state,
    timeline: {
      ...state.timeline,
      holdingScore: state.timeline.holdingScore + holdingPoints,
    },
    markedCells: [...state.markedCells, ...newMarkedCells],
  };
}

/**
 * Clear marked cells and update score
 */
export function clearMarkedCellsAndScore(state: GameState): GameState {
  const { newBoard, newFallingColumns } = clearMarkedCellsAndApplyGravity(
    state.board,
    state.markedCells,
    state.fallingColumns
  );
  const detectedPatterns = detectPatterns(newBoard);

  return {
    ...state,
    board: newBoard,
    fallingColumns: newFallingColumns,
    detectedPatterns,
    score: state.score + state.timeline.holdingScore,
    timeline: {
      ...state.timeline,
      holdingScore: 0,
    },
    markedCells: [],
  };
}

File: python/game/timeline.py
"""
timeline.py — Timeline sweep and scoring logic.
Port of src/utils/gameLogic/timeline.ts
"""

import copy
from .constants import BOARD_WIDTH
from .types import GameState
from .patterns import detect_patterns, get_patterns_by_left_column, mark_column_cells
from .physics import clear_marked_cells_and_apply_gravity


def update_timeline(state: GameState) -> GameState:
    """
    Advance the timeline timer by one frame.
    When the sweep interval is reached, advance the timeline to the next column.
    """
    if state.status != 'playing':
        return state

    new_timer = state.timeline.timer + 1

    if new_timer >= state.timeline.sweep_interval:
        return _advance_timeline_to_next_column(state)

    # Just increment timer
    state = copy.copy(state)
    tl = copy.copy(state.timeline)
    tl.timer = new_timer
    state.timeline = tl
    return state


def _advance_timeline_to_next_column(state: GameState) -> GameState:
    """Move the timeline sweep one column to the right and process it."""
    next_column = (state.timeline.x + 1) % BOARD_WIDTH

    processed = _process_timeline_column(state, next_column)

    processed = copy.copy(processed)
    tl = copy.copy(processed.timeline)
    tl.x = next_column
    tl.timer = 0
    processed.timeline = tl
    return processed


def _process_timeline_column(state: GameState, column: int) -> GameState:
    """Process a single column as the timeline sweeps through it."""
    patterns_in_column = get_patterns_by_left_column(state.detected_patterns, column)
    patterns_in_prev_column = get_patterns_by_left_column(state.detected_patterns, column - 1)

    has_current = len(patterns_in_column) > 0
    has_prev = len(patterns_in_prev_column) > 0

    # Case 1: Mark cells for clearing if there are patterns
    if has_current or has_prev:
        return _mark_cells_for_clearing(state, column, patterns_in_column)

    # Case 2: Clear marked cells if no patterns and we have pending score
    should_clear = (
        not has_current and
        not has_prev and
        state.timeline.holding_score > 0 and
        len(state.marked_cells) > 0
    )

    if should_clear:
        return _clear_marked_cells_and_score(state)

    return state


def _mark_cells_for_clearing(state, column, patterns_in_column):
    """Mark cells for clearing and accumulate holding score."""
    holding_points = len(patterns_in_column)
    new_marked = mark_column_cells(column, state.detected_patterns)

    state = copy.copy(state)
    tl = copy.copy(state.timeline)
    tl.holding_score = tl.holding_score + holding_points
    state.timeline = tl
    state.marked_cells = list(state.marked_cells) + new_marked
    return state


def _clear_marked_cells_and_score(state: GameState) -> GameState:
    """Clear all marked cells, apply gravity, and add holding_score to score."""
    new_falling_cols, new_board = clear_marked_cells_and_apply_gravity(
        state.board,
        state.marked_cells,
        state.falling_columns,
    )
    detected = detect_patterns(new_board)

    state = copy.copy(state)
    state.board = new_board
    state.falling_columns = new_falling_cols
    state.detected_patterns = detected
    state.score = state.score + state.timeline.holding_score
    tl = copy.copy(state.timeline)
    tl.holding_score = 0
    state.timeline = tl
    state.marked_cells = []
    return state
